source/common/router/header_parser.cc

```diff
@@ -286,12 +288,26 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
   for (const auto& [key, entry] : headers_to_add_) {
     const std::string value =
         stream_info != nullptr ? entry.formatter_->format(*stream_info) : entry.original_value_;
-    if (!value.empty()) {
-      if (entry.formatter_->append()) {
-        headers.addReferenceKey(key, value);
-      } else {
+    switch (entry.formatter_->appendAction()) {
+    case envoy::config::core::v3::HeaderValueOption::ADD_IF_ABSENT:
+      // Check whether the header already exist or not. We only
+      // need to add the header if it doesn't already exist.
+      if (headers.get(key).empty()) {
         headers.setReferenceKey(key, value);
```

Comment:
Update this `setReferenceKey` to `addReferenceKey` to reduce unnecessary header map searching.

Reply:
Done!